cadcutils/cadcutils/net/ws.py

```diff
@@ -531,6 +534,186 @@ def upload_file(self, url, src, md5_checksum=None, **kwargs):
                     'File {} not properly uploaded: HTTPStatus {}'.format(
                         src, response.status_code))
 
+    def _resolve_destination_file(self, dest, src_md5, default_file_name):
+        # returns destination absolute file name as well as a temporary
+        # destination to be used during the transfer
+        if (dest is None) and (default_file_name is None):
+            raise ValueError('BUG: Cannot resolve file name')
+        if dest is not None:
+            # got a dest name?
+            if os.path.isdir(dest):
+                final_dest = os.path.join(dest, default_file_name)
+            else:
+                final_dest = dest
+        else:
+            final_dest = os.path.basename(default_file_name)
+        if src_md5:
+            dir_name = os.path.dirname(final_dest)
+            file_name = os.path.basename(final_dest)
+            temp_dest = os.path.join(
+                dir_name, 'cadcget-{}-{}'.format(file_name, src_md5))
+            return final_dest, temp_dest
+        else:
+            return final_dest, final_dest
+
+    def download_file(self, url, dest=None, **kwargs):
+        """Method to download a file from CADC storage (archive or vospace).
+           This method takes advantage of the HTTP Range feature available
+           with the CADC services to optimize and make the transfer more
+           robust.
+           :param url: URL to get the file from
+           :param dest: name of the file to store it to. If it's the name of
+           the directory to save it to, it will use the Content-Disposition for
+           the file name. By default, it saves the file in the current
+           directory.
+           :param kwargs: other http attributes
+           :return: requests.Response object
+           :throws: HttpExceptions
+
+        """
+        response = self.get(url, stream=True)
+        src_md5 = net.extract_md5(response.headers)
+        src_size = int(response.headers.get(HTTP_LENGTH, 0))
+        content_disp = net.get_header_filename(response.headers)
+        if hasattr(dest, 'read'):
+            dest.write(response.raw.read())
+            return
+        else:
+            final_dest, temp_dest = self._resolve_destination_file(
+                dest=dest, src_md5=src_md5, default_file_name=content_disp)
+            if os.path.isfile(final_dest) and \
+               src_size == os.stat(final_dest).st_size and \
+               src_md5 == \
+                    hashlib.md5(open(final_dest, 'rb').read()).hexdigest():
+                # nothing to be done
+                return
+            if src_md5 and src_size and os.path.isfile(temp_dest):
```

Comment:
Does this `isfile` check duplicate the code at l544?

Reply:
l544 checks the destination whereas this line checks the if the corresponding temporary file exists. I hope I understand right your question.